### PR TITLE
ISPN-10322 Create unified interface for initializing commands

### PR DESCRIPTION
**Pull request by ryanemerson**
@danberindei @wburns  This is an attempt to standardise the way in which we initialise our RPC commands to remove the need for the switch statement in CommandsFactoryImpl. This started when I was bored at the airport, so feel free to tear it apart :smile:.

~~In order to avoid creating a sub-interface, e.g. ComandInitialisationContext, the `InitializableCommand#init` method takes in the `ComponentRegistry` and I have cached the components required throughout core to avoid hash lookups. The `Expose*` commits, was just a logical way of exposing sub-components by managers in order to reduce the amount directly expose by the ComponentRegistry. I'm not sure whether this is a good/bad idea.~~

~~I have left support for `ModuleCommandInitializer` as this can be more powerful e.g. hibernate `CacheCommandInitializer`. However, if we go with this approach we should probably all `ModuleCommandExtensions#getModuleCommandInitializer` to return null or provide a default impl that does this.~~

~~I've only update the core module for now, if this is something to pursue then I will update the other modules that utilise `ModuleCommandInitializer`.~~

https://issues.jboss.org/browse/ISPN-10322

### CODE REVIEW

File: core/src/main/java/org/infinispan/commands/functional/AbstractWriteKeyCommand.java [hunk 1]
```diff
@@ -3,35 +3,45 @@
...
 
-public abstract class AbstractWriteKeyCommand<K, V> extends AbstractDataWriteCommand implements FunctionalCommand<K, V> {
+public abstract class AbstractWriteKeyCommand<K, V> extends AbstractDataWriteCommand implements FunctionalCommand<K, V>, InitializableCommand {
 
    Params params;
    ValueMatcher valueMatcher;
    boolean successful = true;
    DataConversion keyDataConversion;
    DataConversion valueDataConversion;
+   Object function;
```
Comment (galderz) [inline core/src/main/java/org/infinispan/commands/functional/AbstractWriteKeyCommand.java]: Subclasses have a stronger type than `Object`, can you improve this?

**Comment by ryanemerson** [comment]
There's still a  lot of failures on this branch, but I don't want to waste time resolving them until the general idea is approved. If not, I'll just close the PR.

File: core/src/main/java/org/infinispan/commands/functional/AbstractWriteKeyCommand.java (hunk 1 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/commands/functional/AbstractWriteKeyCommand.java]: Agreed, probably through templating, as it seems there is no common "root" interface (some are Functions and some are InjectableComponent).

**Comment by tristantarrant** [comment]
Overall I like this approach a lot.

**Comment by wburns** [comment]
Overall I like the approach as well as it keeps the code cleaner. I would love to see the perf difference; I am guessing it is negligible though (since it is a megamorphic call now instead but doesn't have the switch statement).

**Comment by ryanemerson** [comment]
Updated to apply the approach to all of our modules. @danberindei please take a look when you get a chance as I'm sure you'll be more interested then most.

**Comment by ryanemerson** [comment]
run performance tests please

**Comment by diegolovison** [comment]
@ryanemerson blocked by https://github.com/radargun/radargun/pull/603

**Comment by diegolovison** [comment]
run performance tests please

**Comment by danberindei** [comment]
@ryanemerson I would say it's a step in the right direction, but it doesn't go far enough!

I would like to get rid of the separate remote initialization step completely and instead add a method `CacheRpcCommand.invoke(ComponentRegistry)`. We don't need the `isRemote` flag because we already have `getOrigin() != null`. 

For local invocation, the invoker already knows the concrete type, so instead of adding a `LocalInvocation` layer it should call a concrete method that doesn't need the entire component registry and response unwrapping. Besides, `LocalInvocation` is blocking, so it should go away anyway. But I think this can be done in a separate PR.

Currently global commands use `wireDependencies()`, I'd like to get rid of that in the near future and create a `GlobalRpcCommand` interface with an `invoke(GlobalComponentRegistry)` method.

Further out, I hope to replace the interface `invoke` with an annotation, so commands can declare the components they need as parameters, and generated code takes care of both injection and caching, then I can finally retire `ComponentRegistry` :) It will require one hash map lookup to get the accessor, but I think it's worth it.

**Comment by danberindei** [comment]
Integrated, thanks Ryan!

**Pull request merged**